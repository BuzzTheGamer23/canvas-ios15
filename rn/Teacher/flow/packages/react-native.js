// @flow
export type PanResponderGestureState = {
  stateID: number,
  moveX: number,
  moveY: number,
  x0: number,
  y0: number,
  dx: number,
  dy: number,
  vx: number,
  vy: number,
}
