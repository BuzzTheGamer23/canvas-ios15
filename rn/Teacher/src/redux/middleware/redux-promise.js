// @flow

import { isFSA } from 'flux-standard-action'
import type { Middleware, MiddlewareAPI, Dispatch } from 'redux'

export type Action = {
  payload: any,
  pending?: boolean,
  error?: boolean,
}

function isPromise (val: any): boolean {
  return val && typeof val.then === 'function'
}

/*
 * Simple middleware heavily based of off redux-promise for handling FSA
 * promises:
 *
 *   - When a FSA action is received with a payload that is a promise, or with
 *     a payload with a `promise` member, the action is tagged with a
 *     `pending: true` field.
 *   - When the attached promise resolves, a copy of the action is re-dispatched
 *     with its `payload` set to the promised value.
 *   - When the attached promise rejects, a copy of the action is re-dispatched
 *     with its `payload` set to the error, and the `error: true` tag.
 *
 */
let promiseMiddleware = <S, A: Action>({ dispatch }: MiddlewareAPI<S, A>): Middleware<S, A> => {
  return (next: Dispatch<A>) => (action: A) => {
    let promise = action.payload && action.payload.promise
    if (isFSA(action) && isPromise(promise) && !action.error) {
      promise.then(
        result => {
          let payload = { ...action.payload, result }
          delete payload.promise
          return dispatch({ ...action, payload })
        },
        error => {
          let payload = { ...action.payload, error }
          return dispatch({ ...action, payload, error: true })
        }
      )
      next({
        ...action,
        pending: true,
      })
    } else {
      next(action)
    }
  }
}

export default promiseMiddleware
