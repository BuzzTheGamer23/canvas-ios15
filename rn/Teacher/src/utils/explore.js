/* @flow */

function hasProp (object: any, key: string, value: any): boolean {
  if (object instanceof Object) {
    const props = object.props
    if (props && props instanceof Object) {
      for (var prop in props) {
        if (prop === key && props[prop] === value) {
          return true
        }
      }
    }
  }

  return false
}

function selectByProp (object: any, propKey: string, value: any): ?any {
  if (!(typeof object === 'object')) {
    return null
  }

  if (hasProp(object, propKey, value)) {
    return object
  }

  let result = null

  for (var key in object) {
    if (object[key] instanceof Array) {
      object[key].forEach((e) => {
        result = selectByProp(e, propKey, value)
        if (result) {
          return result
        }
      })
    }

    result = selectByProp(object[key], propKey, value)
    if (result) {
      return result
    }
  }

  return result
}

class ComponentExplorer {
  component: any

  constructor (component: any) {
    this.component = component
  }

  selectByProp (propKey: string, value: any): ?any {
    return selectByProp(this.component, propKey, value)
  }

  selectByID (id: string): ?any {
    return this.selectByProp('testID', id)
  }
}

export default function explore (component: any): ComponentExplorer {
  return new ComponentExplorer(component)
}
