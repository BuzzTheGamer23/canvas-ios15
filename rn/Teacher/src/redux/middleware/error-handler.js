//
// Copyright (C) 2016-present Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, version 3 of the License.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
//

// @flow

import { Alert } from 'react-native'
import type { MiddlewareAPI } from 'redux'
import i18n from 'format-message'
import loginVerify from '../../common/login-verify'
export const ERROR_TITLE: string = i18n('Unexpected Error')
export const ERROR_MESSAGE: string = i18n('There was an unexpected error. Please try again.')

const showGlobalErrorAlertIfNecessary = (error: any) => {
  // should only continue if:
  // 1. error is not an api response error
  // 2. error is an api response error and the status code is 500 or above and the user is online
  let isOfflineError = error.message === 'Network Error'
  if (error.response != null) {
    error = error.response
  }
  if (!isOfflineError) {
    let errorMessage = parseErrorMessage(error)
    Alert.alert(ERROR_TITLE, errorMessage)
  }
}

const errorHandlerMiddleware: MiddlewareAPI = () => {
  return next => action => {
    if (action.error) {
      let error = action.payload.error
      if (error && error.response && error.response.status === 401) {
        loginVerify().then((invalid) => {
          if (!invalid) {
            if (action.payload.handlesError) return next(action)
            showGlobalErrorAlertIfNecessary(error)
          }
          next(action)
        })
        return // Bail out of this because of the promise. Next will be called after the promise has finished
      } else {
        if (action.payload.handlesError) return next(action)
        showGlobalErrorAlertIfNecessary(error)
      }
    }
    next(action)
  }
}

export default errorHandlerMiddleware

export function parseErrorMessage (error: any): string {
  if (error instanceof Error) {
    return error.message
  }

  if (error && error.data && error.data.errors && error.data.errors.length > 0) {
    return error.data.errors
        .map(error => error.message)
        .map(message => message.replace(/\.+$/, ''))
        .join('. ')
  } else if (error && error.data && error.data.errors instanceof Object && Object.keys(error.data.errors).length > 0) {
    let data = error.data.errors
    let result = Object.keys(data)
      .map((key, index) => data[key])
      .map(obj => obj.length > 0 ? obj[0] : {})
      .map(obj => `${obj.attribute || ''} ${obj.message || ''}`)
      .join('. ')

    if (result.trim()) {
      return result
    }
  }

  return ERROR_MESSAGE
}
