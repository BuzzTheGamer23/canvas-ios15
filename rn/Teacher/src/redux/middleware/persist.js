//
// Copyright (C) 2016-present Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, version 3 of the License.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
//

// @flow
import { AsyncStorage } from 'react-native'
import { getSession } from 'canvas-api'
import type { MiddlewareAPI } from 'redux'
import hydrate from '../hydrate-action'

const STORE_VERSION = '3'

function storeName (session: Session): string {
  return `redux.state.${session.user.id}.${STORE_VERSION}`
}

async function removeOldStates (session: Session) {
  await AsyncStorage.multiRemove(
    (await AsyncStorage.getAllKeys())
      .filter(k => k.startsWith(`redux.state.${session.user.id}`))
  )
}

async function persistState (store) {
  let state = store.getState()
  let session = getSession()
  if (!session) return

  let expires = new Date()
  expires.setDate(expires.getDate() + 1)
  let cache = {
    expires: expires,
    state,
  }

  await AsyncStorage.setItem(storeName(session), JSON.stringify(cache))
}

export async function hydrateStoreFromPersistedState (store: any): any {
  let session = getSession()
  if (session) {
    let cachedState = await AsyncStorage.getItem(storeName(session))
    if (!cachedState) {
      await removeOldStates(session)
    }
    let appState
    try {
      appState = JSON.parse(cachedState)
    } catch (err) {}
    store.dispatch(hydrate(appState))
  }
}

const createPersistMiddleware = (timeoutWait: number): MiddlewareAPI => {
  return (store) => {
    let timeout
    return next => async (action) => {
      next(action)

      clearTimeout(timeout)
      timeout = setTimeout(() => persistState(store), timeoutWait)
    }
  }
}

export default createPersistMiddleware
