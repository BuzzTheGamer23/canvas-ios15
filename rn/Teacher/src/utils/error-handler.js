// @flow

import { Alert } from 'react-native'
import type { MiddlewareAPI } from 'redux'
import i18n from 'format-message'

export const ERROR_TITLE = i18n({
  default: 'Unexpected Error',
  description: 'The generic title of the generic error message',
})
export const ERROR_MESSAGE = i18n({
  default: 'There was an unexpected error. Please close this alert and try again.',
  description: 'A generic error message',
})

const errorHandlerMiddleware: MiddlewareAPI = () => {
  return next => action => {
    if (action.error) {
      // should only continue if:
      // 1. error is not an api response error
      // 2. error is an api response error and the status code is 500 or above
      // 3. the action hasn't indicated it will handle the error itself
      let error = action.payload.error || action.payload
      let isApiError = error.data && error.status
      if (!isApiError || error.status >= 500 || !action.payload.handlesError) {
        let errorMessage = error.message || ERROR_MESSAGE
        if (error.data && error.data.errors && error.data.errors.length > 0) {
          errorMessage = error.data.errors[0].message
        }
        Alert.alert(ERROR_TITLE, errorMessage)
      }
    }
    next(action)
  }
}

export default errorHandlerMiddleware
