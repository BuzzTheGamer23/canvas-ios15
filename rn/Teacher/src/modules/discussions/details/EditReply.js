//
// Copyright (C) 2016-present Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, version 3 of the License.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
//

// @flow

import React, { Component } from 'react'
import { connect } from 'react-redux'
import i18n from 'format-message'
import color from '../../../common/colors'
import Screen from '../../../routing/Screen'
import RichTextEditor from '../../../common/components/rich-text-editor/RichTextEditor'
import Actions from './actions'
import { ERROR_TITLE } from '../../../redux/middleware/error-handler'
import ModalActivityIndicator from '../../../common/components/ModalActivityIndicator'
import {
    View,
    LayoutAnimation,
    Alert,
} from 'react-native'

type OwnProps = {
  discussionID: string,
  courseID: string,
  indexPath?: number[],
  entryID?: ?string,
  lastReplyAt: string,
}

type State = {
  pending: number,
  error?: ?string,
}

type Props = OwnProps & typeof Actions & NavigationProps & State

export class EditReply extends React.Component<any, Props, any> {
  props: Props

  constructor (props: Props) {
    super(props)
    this.state = {
      pending: false,
    }
  }

  render () {
    let message = this.props.message || ''
    return (
      <Screen
        title={i18n('Reply')}
        navBarStyle='light'
        navBarTitleColor={color.darkText}
        navBarButtonColor={color.link}
        rightBarButtons={[
          {
            title: i18n('Done'),
            style: 'done',
            testID: 'edit-discussion-reply.done-btn',
            action: this._actionDonePressed,
          },
        ]}
        leftBarButtons={[
          {
            title: i18n('Cancel'),
            testID: 'edit-discussion-reply.cancel-btn',
            action: this._actionCancelPressed,
          },
        ]}
      >
        <View style={{ flex: 1 }}>
          <ModalActivityIndicator text={i18n('Saving')} visible={this.state.pending} />
          <RichTextEditor
            onChangeValue={this._valueChanged('message')}
            defaultValue={message}
            showToolbar='always'
            scrollEnabled={true}
            placeholder={i18n('Message')}
            focusOnLoad={true}
          />
        </View>
      </Screen>
    )
  }

  componentWillReceiveProps (props: Props) {
    if (props.error) {
      this.setState({ pending: false })
      this._handleError(props.error)
      return
    }
    if (this.state.pending && !props.pending) {
      this.setState({ pending: false })
      this.props.refreshDiscussionEntries(this.props.courseID, this.props.discussionID, true)
      this.props.navigator.dismissAllModals()
      return
    }
  }

  componentWillUnmount () {
    this.props.deletePendingReplies(this.props.discussionID)
  }

  _actionDonePressed = () => {
    const params = {
      message: this.state.message,
    }
    this.setState({ pending: true })
    if (this.props.isEdit) {
      this.props.editEntry(this.props.courseID, this.props.discussionID, this.props.entryID, params, this.props.indexPath)
    } else {
      this.props.createEntry(this.props.courseID, this.props.discussionID, this.props.entryID, params, this.props.indexPath, this.props.lastReplyAt)
    }
  }

  _actionCancelPressed = () => {
    this.props.navigator.dismiss()
  }

  _valueChanged (property: string, transformer?: any, animated?: boolean = true): Function {
    return (value) => {
      if (transformer) { value = transformer(value) }
      this._valuesChanged({ [property]: value }, animated)
    }
  }

  _valuesChanged (values: Object, animated?: boolean) {
    if (animated) {
      LayoutAnimation.easeInEaseOut()
    }
    this.setState({ ...values })
  }

  _handleError (error: string) {
    setTimeout(() => {
      Alert.alert(ERROR_TITLE, error)
    }, 1000)
  }
}

export function mapStateToProps ({ entities }: AppState, { courseID, discussionID }: OwnProps): State {
  let pending = 0
  let error = null

  if (entities.discussions[discussionID] &&
      entities.discussions[discussionID].replies &&
      entities.discussions[discussionID].replies.new) {
    pending = entities.discussions[discussionID].replies.new.pending
    error = entities.discussions[discussionID].replies.new.error
  }
  if (!pending && !error &&
      entities.discussions[discussionID] &&
      entities.discussions[discussionID].replies &&
      entities.discussions[discussionID].replies.edit) {
    pending = entities.discussions[discussionID].replies.edit.pending
    error = entities.discussions[discussionID].replies.edit.error
  }

  return {
    pending,
    error,
  }
}

let Connected = connect(mapStateToProps, Actions)(EditReply)
export default (Connected: Component<any, Props, any>)
