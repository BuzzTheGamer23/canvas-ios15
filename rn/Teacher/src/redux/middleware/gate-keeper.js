//
// Copyright (C) 2016-present Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, version 3 of the License.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
//

// @flow

import type { MiddlewareAPI } from 'redux'
import { Alert } from 'react-native'
import { HYDRATE_ACTION } from '../hydrate-action'
import logoutAction from '../logout-action'
import { isFSA } from 'flux-standard-action'
import i18n from 'format-message'
import { isOnline } from '../../utils/online-status'

let isLoggedIn = false
let isAlertOpen = false

export function updateAlertState (state: boolean): void {
  isAlertOpen = state
}

const gateKeeperMiddleware: MiddlewareAPI = () => {
  return next => action => {
    if (action.type === HYDRATE_ACTION) {
      isLoggedIn = true
    }

    if (action.type === logoutAction.type) {
      isLoggedIn = false
    }

    let promise = action.payload && action.payload.promise || action.payload
    if (!isOnline() && isFSA(action) && promise instanceof Promise) {
      if (!isAlertOpen) {
        updateAlertState(true)
        Alert.alert(
          i18n('No internet connection'),
          i18n('This action requires an internet connection.'),
          [
            { text: 'Cancel', onPress: () => updateAlertState(false), style: 'cancel' },
          ]
        )
      }
    }

    if (isLoggedIn || action.type === logoutAction.type) {
      next(action)
    }
  }
}

export default gateKeeperMiddleware
