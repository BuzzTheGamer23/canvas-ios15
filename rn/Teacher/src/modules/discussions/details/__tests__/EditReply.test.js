//
// Copyright (C) 2016-present Instructure, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

// @flow

import React from 'react'
import { EditReply, mapStateToProps } from '../EditReply'
import explore from '../../../../../test/helpers/explore'
import setProps from '../../../../../test/helpers/setProps'
import { Alert } from 'react-native'
import { defaultErrorTitle } from '../../../../redux/middleware/error-handler'
import renderer from 'react-test-renderer'

jest
  .mock('Alert', () => ({
    alert: jest.fn(),
  }))
  .mock('../../../../routing')
  .mock('../../../../routing/Screen')
  .mock('../../../../common/components/rich-text-editor/RichTextEditor', () => 'RichTextEditor')
  .mock('LayoutAnimation', () => ({
    easeInEaseOut: jest.fn(),
    Types: {
      easeInEaseOut: jest.fn(),
      spring: jest.fn(),
    },
    Properties: {
      opacity: 1,
    },
  }))

const template = {
  ...require('../../../../__templates__/discussion'),
  ...require('../../../../__templates__/course'),
  ...require('../../../../__templates__/users'),
  ...require('../../../../__templates__/helm'),
  ...require('../../../../redux/__templates__/app-state'),
}

describe('EditReply', () => {
  let defaultProps
  beforeEach(() => {
    jest.resetAllMocks()
    defaultProps = {
      discussion: template.discussion({ id: '1' }),
      navigator: template.navigator(),
      discussionID: '1',
      courseID: '1',
      entryID: '1',
      course: template.course({ id: 1 }),
      indexPath: [],
      deletePendingReplies: jest.fn(),
      lastReplyAt: (new Date()).toISOString(),
    }
  })

  it('renders', () => {
    let tree = renderer.create(
      <EditReply {...defaultProps} />
    ).toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('presents error alert', () => {
    jest.useFakeTimers()
    let component = renderer.create(
      <EditReply {...defaultProps} />
    )

    let errorMessage = 'error'

    setProps(component, { error: errorMessage })
    jest.runAllTimers()

    expect(Alert.alert).toHaveBeenCalledWith(defaultErrorTitle(), errorMessage)
  })

  it('calls dismiss on cancel', () => {
    let component = renderer.create(
      <EditReply {...defaultProps} />
    )
    let postReply = jest.fn(() => {
      setProps(component, { pending: 0, error: 'error' })
    })
    component.update(<EditReply {...defaultProps} updateCourse={postReply} />)
    const doneButton: any = explore(component.toJSON()).selectLeftBarButton('edit-discussion-reply.cancel-btn')
    doneButton.action()
    expect(component.toJSON()).toMatchSnapshot()
  })

  it('deletes pending replies on unmount', () => {
    let component = renderer.create(
      <EditReply {...defaultProps} />
    )
    component.getInstance().componentWillUnmount()
    expect(defaultProps.deletePendingReplies).toHaveBeenCalledWith(defaultProps.discussionID)
  })

  it('dismisses modal activity upon save error', () => {
    let component = renderer.create(
      <EditReply {...defaultProps} />
    )
    let postReply = jest.fn(() => {
      setProps(component, { pending: 0, error: 'error' })
    })
    component.update(<EditReply {...defaultProps} createEntry={postReply} />)
    const doneButton: any = explore(component.toJSON()).selectRightBarButton('edit-discussion-reply.done-btn')
    doneButton.action()
    expect(component.toJSON()).toMatchSnapshot()
  })

  it('dismisses modal after reply updates', () => {
    let component = renderer.create(
      <EditReply {...defaultProps} />
    )
    let postReply = jest.fn(() => {
      setProps(component, { pending: 0 })
    })
    let refresh = jest.fn()
    component.update(<EditReply {...defaultProps} createEntry={postReply} refreshDiscussionEntries={refresh} />)
    const doneButton: any = explore(component.toJSON()).selectRightBarButton('edit-discussion-reply.done-btn')
    doneButton.action()
    expect(defaultProps.navigator.dismissAllModals).toHaveBeenCalled()
  })

  it('sets message placeholder', () => {
    let component = renderer.create(
      <EditReply {...defaultProps} />
    )
    let textEditor = explore(component.toJSON()).query(({ type }) => type === 'RichTextEditor')[0]

    expect(textEditor.props.placeholder).toEqual('Message')
  })

  it('enteres text and posts reply', () => {
    let component = renderer.create(
      <EditReply {...defaultProps} />
    )
    let postReply = jest.fn(() => {
      setProps(component, { pending: 0 })
    })
    let textEditor = explore(component.toJSON()).query(({ type }) => type === 'RichTextEditor')[0]
    let message = 'not empty'
    textEditor.props.onChangeValue(message)
    let refresh = jest.fn()
    component.update(<EditReply {...defaultProps} createEntry={postReply} refreshDiscussionEntries={refresh} />)
    const doneButton: any = explore(component.toJSON()).selectRightBarButton('edit-discussion-reply.done-btn')
    doneButton.action()
    expect(postReply).toBeCalledWith(defaultProps.courseID, defaultProps.discussionID, defaultProps.entryID, { message }, [], defaultProps.lastReplyAt)
    expect(defaultProps.navigator.dismissAllModals).toHaveBeenCalled()
  })

  it('edits an existing reply', () => {
    let editProps = {
      ...defaultProps,
      message: 'default message',
      isEdit: true,
    }
    let component = renderer.create(
      <EditReply {...editProps} />
    )
    let editReply = jest.fn(() => {
      setProps(component, { pending: 0 })
    })
    let textEditor = explore(component.toJSON()).query(({ type }) => type === 'RichTextEditor')[0]
    let message = 'edited message'
    textEditor.props.onChangeValue(message)
    let refresh = jest.fn()
    component.update(<EditReply {...editProps} editEntry={editReply} refreshDiscussionEntries={refresh} />)
    const doneButton: any = explore(component.toJSON()).selectRightBarButton('edit-discussion-reply.done-btn')
    doneButton.action()
    expect(editReply).toBeCalledWith(editProps.courseID, editProps.discussionID, editProps.entryID, { message }, [])
    expect(defaultProps.navigator.dismissAllModals).toHaveBeenCalled()
  })
})

describe('MapStateToProps', () => {
  let lastReplyAt = (new Date(0)).toISOString()
  test('maps default data for new reply', () => {
    const state: AppState = template.appState({
      entities: {
        discussions: {

        },
      },
    })
    expect(
      mapStateToProps(state, { courseID: '1', discussionID: '1', indexPath: [], lastReplyAt })
    ).toMatchObject({
      pending: 0,
      error: null,
    })
  })

  test('finds the correct data for new reply', () => {
    const state: AppState = template.appState({
      entities: {
        ...template.appState().entities,
        discussions: {
          '1': {
            pending: 0,
            error: null,
            replies: {
              pending: 0,
              error: null,
              refs: [],
              new: {
                pending: 14,
                error: 'Map this error',
              },
            },
          },
        },
      },
    })
    expect(
      mapStateToProps(state, { courseID: '1', discussionID: '1', lastReplyAt })
    ).toMatchObject({
      pending: 14,
      error: 'Map this error',
    })
  })

  test('finds the correct data for editing a reply', () => {
    const state: AppState = template.appState({
      entities: {
        ...template.appState().entities,
        discussions: {
          '1': {
            pending: 0,
            error: null,
            replies: {
              pending: 0,
              error: null,
              refs: [],
              edit: {
                pending: 14,
                error: 'Map this error',
              },
            },
          },
        },
      },
    })
    expect(
      mapStateToProps(state, { courseID: '1', discussionID: '1', lastReplyAt })
    ).toMatchObject({
      pending: 14,
      error: 'Map this error',
    })
  })
})
