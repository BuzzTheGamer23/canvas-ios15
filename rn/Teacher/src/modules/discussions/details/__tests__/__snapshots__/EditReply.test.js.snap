// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditReply calls dismiss on cancel 1`] = `
<Screen
  leftBarButtons={
    Array [
      Object {
        "action": [Function],
        "testID": "edit-discussion-reply.cancel-btn",
        "title": "Cancel",
      },
    ]
  }
  navBarButtonColor="#008EE2"
  navBarStyle="light"
  navBarTitleColor="#2D3B45"
  rightBarButtons={
    Array [
      Object {
        "action": [Function],
        "style": "done",
        "testID": "edit-discussion-reply.done-btn",
        "title": "Done",
      },
    ]
  }
  title="Reply"
>
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <Modal
      animationType="fade"
      hardwareAccelerated={false}
      supportedOrientations={
        Array [
          "portrait",
          "landscape",
        ]
      }
      transparent={true}
      visible={false}
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "flex": 1,
              "flexDirection": "row",
              "justifyContent": "center",
            },
          ]
        }
      >
        <View
          style={
            Array [
              Object {
                "backgroundColor": "rgba(0,0,0,0.6)",
                "borderRadius": 12,
              },
              Object {
                "backgroundColor": "rgba(0,0,0,0.8)",
              },
              Object {
                "height": 170,
                "width": 170,
              },
            ]
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "flex": 1,
                "flexDirection": "column",
                "justifyContent": "center",
              }
            }
          >
            <View
              style={
                Object {
                  "alignItems": "center",
                  "flex": 1,
                  "flexDirection": "column",
                  "justifyContent": "center",
                }
              }
            >
              <Text
                accessible={true}
                allowFontScaling={true}
                disabled={false}
                ellipsizeMode="tail"
                style={
                  Array [
                    Object {
                      "fontFamily": ".SFUIDisplay",
                    },
                    Object {
                      "color": "#fff",
                      "fontFamily": ".SFUIDisplay-semibold",
                      "fontSize": 24,
                    },
                    Array [
                      Object {
                        "alignItems": "center",
                        "color": "#fff",
                        "textAlign": "center",
                      },
                      undefined,
                    ],
                  ]
                }
              >
                Saving
              </Text>
            </View>
          </View>
          <ActivityIndicator
            animating={true}
            color="#fff"
            hidesWhenStopped={true}
            size="large"
            style={
              Object {
                "flex": 1,
                "marginTop": -40,
              }
            }
          />
        </View>
      </View>
    </Modal>
    <RichTextEditor
      defaultValue=""
      focusOnLoad={true}
      onChangeValue={[Function]}
      placeholder="Message"
      scrollEnabled={true}
      showToolbar="always"
    />
  </View>
</Screen>
`;

exports[`EditReply dismisses modal activity upon save error 1`] = `
<Screen
  leftBarButtons={
    Array [
      Object {
        "action": [Function],
        "testID": "edit-discussion-reply.cancel-btn",
        "title": "Cancel",
      },
    ]
  }
  navBarButtonColor="#008EE2"
  navBarStyle="light"
  navBarTitleColor="#2D3B45"
  rightBarButtons={
    Array [
      Object {
        "action": [Function],
        "style": "done",
        "testID": "edit-discussion-reply.done-btn",
        "title": "Done",
      },
    ]
  }
  title="Reply"
>
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <Modal
      animationType="fade"
      hardwareAccelerated={false}
      supportedOrientations={
        Array [
          "portrait",
          "landscape",
        ]
      }
      transparent={true}
      visible={false}
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "flex": 1,
              "flexDirection": "row",
              "justifyContent": "center",
            },
          ]
        }
      >
        <View
          style={
            Array [
              Object {
                "backgroundColor": "rgba(0,0,0,0.6)",
                "borderRadius": 12,
              },
              Object {
                "backgroundColor": "rgba(0,0,0,0.8)",
              },
              Object {
                "height": 170,
                "width": 170,
              },
            ]
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "flex": 1,
                "flexDirection": "column",
                "justifyContent": "center",
              }
            }
          >
            <View
              style={
                Object {
                  "alignItems": "center",
                  "flex": 1,
                  "flexDirection": "column",
                  "justifyContent": "center",
                }
              }
            >
              <Text
                accessible={true}
                allowFontScaling={true}
                disabled={false}
                ellipsizeMode="tail"
                style={
                  Array [
                    Object {
                      "fontFamily": ".SFUIDisplay",
                    },
                    Object {
                      "color": "#fff",
                      "fontFamily": ".SFUIDisplay-semibold",
                      "fontSize": 24,
                    },
                    Array [
                      Object {
                        "alignItems": "center",
                        "color": "#fff",
                        "textAlign": "center",
                      },
                      undefined,
                    ],
                  ]
                }
              >
                Saving
              </Text>
            </View>
          </View>
          <ActivityIndicator
            animating={true}
            color="#fff"
            hidesWhenStopped={true}
            size="large"
            style={
              Object {
                "flex": 1,
                "marginTop": -40,
              }
            }
          />
        </View>
      </View>
    </Modal>
    <RichTextEditor
      defaultValue=""
      focusOnLoad={true}
      onChangeValue={[Function]}
      placeholder="Message"
      scrollEnabled={true}
      showToolbar="always"
    />
  </View>
</Screen>
`;

exports[`EditReply renders 1`] = `
<Screen
  leftBarButtons={
    Array [
      Object {
        "action": [Function],
        "testID": "edit-discussion-reply.cancel-btn",
        "title": "Cancel",
      },
    ]
  }
  navBarButtonColor="#008EE2"
  navBarStyle="light"
  navBarTitleColor="#2D3B45"
  rightBarButtons={
    Array [
      Object {
        "action": [Function],
        "style": "done",
        "testID": "edit-discussion-reply.done-btn",
        "title": "Done",
      },
    ]
  }
  title="Reply"
>
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <Modal
      animationType="fade"
      hardwareAccelerated={false}
      supportedOrientations={
        Array [
          "portrait",
          "landscape",
        ]
      }
      transparent={true}
      visible={false}
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "flex": 1,
              "flexDirection": "row",
              "justifyContent": "center",
            },
          ]
        }
      >
        <View
          style={
            Array [
              Object {
                "backgroundColor": "rgba(0,0,0,0.6)",
                "borderRadius": 12,
              },
              Object {
                "backgroundColor": "rgba(0,0,0,0.8)",
              },
              Object {
                "height": 170,
                "width": 170,
              },
            ]
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "flex": 1,
                "flexDirection": "column",
                "justifyContent": "center",
              }
            }
          >
            <View
              style={
                Object {
                  "alignItems": "center",
                  "flex": 1,
                  "flexDirection": "column",
                  "justifyContent": "center",
                }
              }
            >
              <Text
                accessible={true}
                allowFontScaling={true}
                disabled={false}
                ellipsizeMode="tail"
                style={
                  Array [
                    Object {
                      "fontFamily": ".SFUIDisplay",
                    },
                    Object {
                      "color": "#fff",
                      "fontFamily": ".SFUIDisplay-semibold",
                      "fontSize": 24,
                    },
                    Array [
                      Object {
                        "alignItems": "center",
                        "color": "#fff",
                        "textAlign": "center",
                      },
                      undefined,
                    ],
                  ]
                }
              >
                Saving
              </Text>
            </View>
          </View>
          <ActivityIndicator
            animating={true}
            color="#fff"
            hidesWhenStopped={true}
            size="large"
            style={
              Object {
                "flex": 1,
                "marginTop": -40,
              }
            }
          />
        </View>
      </View>
    </Modal>
    <RichTextEditor
      defaultValue=""
      focusOnLoad={true}
      onChangeValue={[Function]}
      placeholder="Message"
      scrollEnabled={true}
      showToolbar="always"
    />
  </View>
</Screen>
`;
